httpbakery/agent/agent.go

```diff
@@ -64,81 +63,43 @@ unmarshals to an httpbakery.Error.
 type agentLogin struct {
 	Username  string            `json:"username"`
 	PublicKey *bakery.PublicKey `json:"public_key"`
 }
 
-// agentResponse contains the response to an agent login attempt.
-type agentResponse struct {
-	AgentLogin bool `json:"agent_login"`
+type agent struct {
+	url      *url.URL
+	username string
+	key      *bakery.KeyPair
 }
 
-// ErrNoAgentLoginCookie is the error returned when the expected
-// agent login cookie has not been found.
-var ErrNoAgentLoginCookie = errgo.New("no agent-login cookie found")
-
-// LoginCookie returns details of the agent login cookie
-// from the given request. If no agent-login cookie is found,
-// it returns an ErrNoAgentLoginCookie error.
-func LoginCookie(req *http.Request) (username string, key *bakery.PublicKey, err error) {
-	c, err := req.Cookie(cookieName)
-	if err != nil {
-		return "", nil, ErrNoAgentLoginCookie
+func (a agent) cookie() (*http.Cookie, error) {
+	al := agentLogin{
+		Username:  a.username,
+		PublicKey: &a.key.Public,
 	}
-	b, err := base64.StdEncoding.DecodeString(c.Value)
+	data, err := json.Marshal(al)
 	if err != nil {
-		return "", nil, errgo.Notef(err, "cannot decode cookie value")
+		return nil, errgo.Mask(err)
 	}
-	var al agentLogin
-	if err := json.Unmarshal(b, &al); err != nil {
-		return "", nil, errgo.Notef(err, "cannot unmarshal agent login")
-	}
-	if al.Username == "" {
-		return "", nil, errgo.Newf("agent login has no user name")
-	}
-	if al.PublicKey == nil {
-		return "", nil, errgo.Newf("agent login has no public key")
-	}
-	return al.Username, al.PublicKey, nil
+	return &http.Cookie{
+		Name:  cookieName,
+		Value: base64.StdEncoding.EncodeToString(data),
+	}, nil
 }
 
-// SetUpAuth configures agent authentication on c. A cookie is created in
-// c's cookie jar containing credentials derived from the username and
-// c.Key. c.VisitWebPage is set to VisitWebPage(c). The return is
-// non-nil only if c.Key is nil.
-func SetUpAuth(c *httpbakery.Client, u *url.URL, username string) error {
-	if c.Key == nil {
-		return errgo.New("cannot set-up authentication: client key not configured")
-	}
-	SetCookie(c.Jar, u, username, &c.Key.Public)
-	c.VisitWebPage = VisitWebPage(c)
-	return nil
+type Visitor struct {
+	agents []agent
 }
 
-// SetCookie creates a cookie in jar which is suitable for performing agent
-// logins to u.
-//
-// If using SetUpAuth, it should not be necessary to use
-// this function.
-func SetCookie(jar http.CookieJar, u *url.URL, username string, pk *bakery.PublicKey) {
-	al := agentLogin{
-		Username:  username,
-		PublicKey: pk,
-	}
-	b, err := json.Marshal(al)
-	if err != nil {
-		// This shouldn't happen as the agentLogin type has to be marshalable.
-		panic(errgo.Notef(err, "cannot marshal cookie"))
-	}
-	v := base64.StdEncoding.EncodeToString(b)
-	jar.SetCookies(u, []*http.Cookie{{
-		Name:  cookieName,
-		Value: v,
-	}})
+func (v *Visitor) AddAgent(u *url.URL, username string, key *bakery.KeyPair) {
```

Comment:
Doc comment please.

What are the intended semantics if we have two agents both with matching URLs for a request we're making?